File: analysis/plot_runtime_comparison.m
function plot_runtime_comparison(runtime_csv, output_path)
%PLOT_RUNTIME_COMPARISON Save perception runtime bars for all controllers.
%   plot_runtime_comparison(runtime_csv) reads the CSV produced by the day7
%   thesis package and writes results/figures/runtime_perception_comparison.png.

if nargin < 1 || isempty(runtime_csv)
    runtime_csv = fullfile("results", "perception_logs", "day7_thesis_package", "runtime_profile.csv");
end
if nargin < 2 || isempty(output_path)
    output_path = fullfile("results", "figures", "runtime_perception_comparison.png");
end

if ~exist(runtime_csv, "file")
    error("plot_runtime_comparison:MissingCSV", "Runtime CSV not found: %s", runtime_csv);
end

T = readtable(runtime_csv);
controller = string(T.controller);
values = T.mean_inference_time_ms;

out_dir = fileparts(output_path);
if ~exist(out_dir, "dir")
    mkdir(out_dir);
end

fig = figure("Visible", "off", "Color", "w");
b = bar(categorical(controller), values);
b.FaceColor = [0.20 0.45 0.75];
grid on;
ylabel("Mean perception runtime per frame [ms]");
title("Perception Runtime by Controller");
set(gca, "FontName", "Arial", "FontSize", 10);
for k = 1:numel(values)
    text(k, values(k), sprintf("%.1f", values(k)), ...
        "HorizontalAlignment", "center", "VerticalAlignment", "bottom", "FontSize", 9);
end
exportgraphics(fig, output_path, "Resolution", 200);
close(fig);
end

File: analysis/plot_runtime_comparison.py
"""Plot perception runtime per controller as a bar chart.

Reads the runtime CSV produced by the day7 thesis package and writes
results/figures/runtime_perception_comparison.png.
"""

from __future__ import annotations

import argparse
import csv
from pathlib import Path

import matplotlib.pyplot as plt


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--csv",
        default="results/perception_logs/day7_thesis_package/runtime_profile.csv",
    )
    parser.add_argument(
        "--out",
        default="results/figures/runtime_perception_comparison.png",
    )
    args = parser.parse_args()

    controllers, values = [], []
    with open(args.csv, newline="", encoding="utf-8") as f:
        for row in csv.DictReader(f):
            controllers.append(row["controller"])
            values.append(float(row["mean_inference_time_ms"]))

    out_path = Path(args.out)
    out_path.parent.mkdir(parents=True, exist_ok=True)

    fig, ax = plt.subplots(figsize=(7, 4.5), dpi=200)
    bars = ax.bar(controllers, values, color="#3373BF")
    ax.set_ylabel("Mean perception runtime per frame [ms]")
    ax.set_title("Perception Runtime by Controller")
    ax.grid(True, axis="y", linestyle="--", alpha=0.5)
    ax.set_axisbelow(True)
    for bar, v in zip(bars, values):
        ax.text(
            bar.get_x() + bar.get_width() / 2,
            v,
            f"{v:.1f}",
            ha="center",
            va="bottom",
            fontsize=9,
        )
    fig.tight_layout()
    fig.savefig(out_path)
    print(f"wrote {out_path}")


if __name__ == "__main__":
    main()
